content/en/blog/2021/better-external-authz/index.md

```diff
@@ -0,0 +1,545 @@
+---
+title: Better External Authorization
+subtitle: Integrate external authorization system (e.g. OPA, oauth2-proxy, etc.) with Istio using AuthorizationPolicy
+description: AuthorizationPolicy now supports CUSTOM action to delegate the authorization to external system.
+publishdate: 2021-02-09
+attribution: Yangmin Zhu (Google), Craig Box (Google) and Christian Posta (Solo.io)
+keywords: [authorization,access control,opa,oauth2]
+---
+
+## Background
+
+Istio's authorization policy provides access control for services in the mesh. It is fast, powerful and a widely used
+feature. We have made continuous improvements to make it more flexible since its first release in Istio 1.4, including
+the DENY action, exclusion semantics, the X-Forwarded-For header support, nested JWT claim support and more.
+
+These features improve the flexibility of the authorization policy, but there are still many use cases cannot be supported
+with this model, for example:
+
+- You have your own in-house authorization system that cannot be easily migrated to, or replaced by, the authorization
+  policy;
+
+- You want to integrate with a 3rd-party or industry standard solution (e.g. [Open Policy Agent](https://www.openpolicyagent.org/docs/latest/envoy-authorization/)
+  or [`oauth2` proxy](https://github.com/oauth2-proxy/oauth2-proxy)) which may require use of the
+  [low-level Envoy configuration APIs](/docs/reference/config/networking/envoy-filter/) in Istio, or may not be possible
+  at all;
+
+- The authorization policy just lacks some semantics for your use cases;
+
+## Solution
+
+In Istio 1.9, we have implemented extensibility into authorization policy by introducing a [`CUSTOM` action](https://github.com/istio/api/blob/b24ca4b76fdb51e227baf508c6d884eb8bcb51f5/security/v1beta1/authorization_policy.proto#L296),
+which allows you to delegate the access control decision to an external authorization service.
+
+The `CUSTOM` action allows to integrate Istio with an external authorization system that implements its own custom
+authorization logic. The following diagram shows the high level architecture of this integration:
+
+{{< image width="100%" link="./external_authz.svg" caption="External Authorization Architecture" >}}
+
+In configuration time, the user (e.g. mesh admin) configures the authorization policy with `CUSTOM` action to enable the
+external authorization on a proxy (either gateway or sidecar). The user should make sure the external auth service is up
+and running.
```

Comment:
```suggestion
At configuration time, the mesh admin configures an authorization policy with a `CUSTOM` action to enable the
external authorization on a proxy (either gateway or sidecar). The admin should make sure the external auth service is up
and running.
```

How do you make sure the auth service is up and running, and what happens if it is not?

Reply:
The user could send some test requests to their auth service to check if it is ready. If not, the authorization policy will fallback to always deny the request.